google/auth/app_engine.py

```diff
@@ -34,13 +34,30 @@
 
 
 class Signer(object):
-    """Signs messages using the App Engine app identity service.
+    """Signs messages using the App Engine App Identity service.
 
     This can be used in place of :class:`google.auth.crypt.Signer` when
     running in the App Engine standard environment.
+
+    .. warning::
+        The App Identity service signs bytes using Google-managed keys.
+        Because of this it's possible that the key used to sign bytes will
+        change. In some cases this change can occur between successive calls
+        to :attr:`key_id` and :meth:`sign`. This could result in a signature
+        that was signed with a different key than the one indicated by
+        :attr:`key_id`. It's recommended that if you use this in your code
+        that you account for this behavior by building in retry logic.
     """
-    def __init__(self):
-        self.key_id = None
+
+    @property
+    def key_id(self):
+        """Optional[str]: The key ID used to identify this private key.
+
+        .. note::
+           This makes a request to the App Identity service.
+        """
+        key_id, _ = app_identity.sign_blob(b'')
```

Comment:
The problem is that the order of operations is important here, think about this order of events:

1. `jwt.encode` is called.
2. `jwt.encode` uses `signer.key_id` to figure out the `kid` header. `signer` caches `key_id`.
3. `jwt.encode` uses `signer.sign` to sign the header and payload.
4. Time passes, key gets rotated.
5. `jwt.encode` is called again.
6. `jwt.encode` uses `signer.key_id` to figure out the `kid` header. `signer` *uses the cached `key_id`*, which is now wrong.
7. `jwt.encode` uses `signer.sign` to sign the header and payload.
8. A client tries to verify the jwt and fails because the kid doesn't match.